google/blockly
# April 2018 release

### Bugs fixed in this release (not all bugs were present on master):
- #1584 (Proposal for extensible input/field types from the JSON block definition)
- #1613 (Broken tooltip: BKY_TEXT_CREATE_JOIN_ITEM_TOOLTIP)
- #1619 (Var id instead of var name in tooltip)
- #1626 (Test blocks in Playground)
- #1642 (Cannot drag from toolbox a procedure definition with arguments)
- #1649 (jsdoc warnings: @type does not permit a description)
- #1633 (If block definition has invalid color, creating/loading fails/throws)
- #1671 (New field registration code does not work under advanced compilation)
- #1671 (Split events.js by event type, and move into a subdirectory)
- #1655 (Tracking issue for pointer events)
- #1695 (Deprecated lua code is generated)
- #391 (Hook to add items in main context menu)
- #1704 (z-order in custom dialogs demo)
- #1699 (Blockly.Mutator.findParentWs is @package)
- #1726 (un/bindEvent_ is package-private)
- #1484 (Blockly background is always black in latest Edge (41) when there is a non-zero number of blocks present)
- #1295 (Support pinch-to-zoom)
- #1741 (Blocks with variables shouldn't show "create var getter" option in context menu when in flyout)
- #1745 (getAddedVariables fails on IE)
- #1485 (Cannot remove single block from a stack anymore.)
- #1648 (Overriding color constants does not override all blocks with given color)
- #1425 (Delete areas broken in RTL Mutators on FF)
- #1659 (Make math_number a 1 by default.)
- #1658 (Context menu create option shows id on blocks with variables)
- #1408 (logic_compare: Type-verification creates top level shadow block as a side effect)

-------------------
### Release notes:
- Implement #1584 (Proposal for extensible input/field types from the JSON block definition)
- Don’t register mouse events if pointer events are supported
- Fix #1613 (Broken tooltip: BKY_TEXT_CREATE_JOIN_ITEM_TOOLTIP)
- Reduce size of XML by omitting empty variable tag.
- Update Closure to 20180204.0.0
- Delete extra variables at the end of a procedure argument edit
- Adding test blocks into the playground
- Adding an <iframe> wrapper page for testing
- Fix #1619 (Var id instead of var name in tooltip)
- Fix #1626 (Test blocks in Playground)
- BlockFactory: Fixing incorrect <button> html
- Serialize variable IDs in procedure definitions (Fix #1642)
- Adding new playground test category "Drag"
- Update comment on loop types
- Fix some annotations (Fix #1649)
- Adding test blocks for Unicode and Emojis
- Print a warning if block JSON color value is invalid (Fixes #1633)
- Comments on generator test workspace save files
- Random typos.
- Move ordinal in block 'text_getSubstring'
- move events.js into an events folder inside core
- Move event classes into separate files, and update requires accordingly
- Field.Variable's dispose should be on the prototype
- Use @nocompress annotation to fix advanced compilation. (Fixes #1671)
- Don't register mouse events if pointer events are supported
- Pointer Event fixes (Fixes #1655)
- Move away from math.log10 and math.pow in Lua generators. (Fixes #1695)
- Expose block ID in SVG.
- Fix lots of compiler warnings and annotations
- Add hook for custom context menus on the workspace (Fixes #391)
- Custom Dialogs demo: Bumping backdrop Z. (Fixes #1704)
- Add IDs to bubbles.
- Add warning_level directives to build.
- Use dataset instead of ID.
- Don't compute (useless) location of trashcan while workspace is being built
- Prevent fatal errors in browsers that don't support dataset.
- Fix #1699 (Blockly.Mutator.findParentWs is @package)
- Add source code link to custom dialog demo.
- Tighten line length rule from 120 to 100
- Add keyword-spacing rule to eslint
- Make bind/unbindEvent_ public
- Dragging blocks towards the toolbox should create blocks rather than do nothing
- Add @this annotations in blocks/text.js
- Stop storing IDs on every block in our datastore. (In demos)
- Create an empty pattern if there is no grid (Fixes #1484)
- Enable touch gestures and pinch-to-zoom (Fixes #1295)
- Don't show create VAR and create Function context menu options when in the flyout (Fixes #1741)
- Add a test block with a dropdown containing both text and images
- Fixes #1745 (getAddedVariables fails on IE)
- Snap a single block from a stack (if meta key is pressed)
- Allow block xml disable tag to be more accepting
- Strip coordinates from storage of single block stacks. (In demos)
- Add a function to scroll the workspace to center on a given block
- Add argsIgnorePattern to the eslintrc 
- Simple tests for getContentDimensions
- Making all blocks use the MSG colour values. (Fixes #1648)
- Work around a problem with RTL mutators (Fixes #1425)
- Create a script to do local builds of core Blockly. 
- Require goog.userAgent in the correct files.
- Remove rogue line breaks in translated messages.
- Setting the value of math_numnber toolbox blocks to 123. (Fixes #1659)
- Fix context menus on blocks with variables (Fixes #1658)
- Rewrote LOGIC_COMPARE_ONCHANGE_MIXIN to fix #1408
